### PR TITLE
add support for large file hierarchies to file explorer

### PR DESCRIPTION
**Pull request by serapath**

### CODE REVIEW

File: src/app/files/basicReadOnlyExplorer.js [hunk 1]
```diff
@@ -81,7 +81,18 @@ class BasicReadOnlyExplorer {
...
+    var tree = {}
+    // This does not include '.remix.config', because it is filtered
+    // inside list().
+    Object.keys(this.list()).forEach(function (path) {
+      hashmapize(tree, path, {
+        '/readonly': self.isReadOnly(path),
+        '/content': self.get(path)
+      })
+    })
+    return callback(null, tree[path] || {})
```
Comment (yann300) [inline src/app/files/basicReadOnlyExplorer.js]: return `tree[path]` always undefined

File: src/app/files/file-explorer.js [hunk 2]
```diff
@@ -99,8 +100,9 @@ function fileExplorer (appAPI, files) {
         })) : undefined
       }
     },
-    formatSelf: function (key, data, li) {
-      var isRoot = data.path.indexOf('/') === -1
+    formatSelf: function formatSelf (key, data, li) {
+      var isRoot = !data.path.indexOf('browser') && data.path.length === 'browser'.length
```
Comment (yann300) [inline src/app/files/file-explorer.js]: why `var isRoot = data.path.indexOf('/') === -1` is not more applicable? I think it still is

File: src/app/files/file-explorer.js [hunk 3]
```diff
@@ -138,10 +162,9 @@ function fileExplorer (appAPI, files) {
   var textUnderEdit = null
   var textInRename = false
 
-  var events = new EventManager()
-  this.events = events
-  var api = {}
-  api.addFile = function addFile (file) {
+  self.events = new EventManager()
+  self.api = {}
+  self.api.addFile = function addFile (file) {
```
Comment (yann300) [inline src/app/files/file-explorer.js]: the `addFile`  button seems to be broken

File: src/app/files/basicReadOnlyExplorer.js (hunk 1 above)
Comment (serapath) [reply src/app/files/basicReadOnlyExplorer.js]: it's not always undefined.
sometimes it return `{ localhost: { foo: {}, bar: {}, baz: {} }`

...or ...hmm, maybe not? I think without it, the names like `gist`, `github`, etc... showed up in the file explorer

File: src/app/files/file-explorer.js (hunk 2 above)
Comment (serapath) [reply src/app/files/file-explorer.js]: hmm... maybe.
I think i had  scenario where this wasnt working, but don't remember what it was.

:/

File: src/app/files/file-explorer.js (hunk 3 above)
Comment (serapath) [reply src/app/files/file-explorer.js]: hm - i didn't touch the `addFile` button, but it's probably related to emitting an event - and i think i added `@TODO` tags to those - so yes that needs fixing

File: src/app/files/basicReadOnlyExplorer.js (hunk 1 above)
Comment (yann300) [reply src/app/files/basicReadOnlyExplorer.js]: well when I tested it with remixd it would always return `undefined`.
use publish to swarm to reproduce

Comment (yann300) [reply src/app/files/basicReadOnlyExplorer.js]: ```I think without it, the names like gist, github, etc... showed up in the file explorer```
that's ok

File: src/app/files/file-explorer.js (hunk 3 above)
Comment (yann300) [reply src/app/files/file-explorer.js]: Let's fix in the next PR

File: src/app/files/file-explorer.js (hunk 2 above)
Comment (yann300) [reply src/app/files/file-explorer.js]: could you rollback for now?

**Comment by yann300** [comment]
the Ballot.sol default file is not loaded by default anymore...

**Comment by serapath** [comment]
oh really? 
hmm, it loaded for me

**Comment by wowa1983** [comment]
Kann mich nicht einloggen zu github habe neue Teil.nummer wie kann ich
Tellefon zurück setzen athetivikator

Am 15.01.2018 18:38 schrieb "Alexander Praetorius" <notifications@github.com

File: src/app/files/file-explorer.js (hunk 2 above)
Comment (serapath) [reply src/app/files/file-explorer.js]: ok

**Pull request merged**